### PR TITLE
Added stage 2 of the registration wizard using the aurelia-validation…

### PR DESCRIPTION
**Pull request by IonelLupu**
## What was done
The stage 2 of the registration wizard was added. Used the aurelia-validation package for form validation.

Preview: https://prime-deals-dapp-git-feature-registration-wiza-20db25-curvelabs.vercel.app/initiate/token-swap/open-proposal/proposal-lead

The form validation is automatically done by using the (VERY IMPORTANT)  ` this.form.validateTrigger  = "changeOrBlur" `.

I also added a custom ValidationController renderer that can be used to automatically display errors in `pform-input` when the `pform-input` has only one element of type "input". I will use this in the next stages of the wizard or any other form.

It means there are two ways of displaying errors: 
1. you listen to the validation event, collected the errors and manually pass them to `pform-input` (used in this PR).
2. you just add the renderer to the validation form (`form.addRenderer(new PrimeRenderer)`) and let it add the errors on your `pform-inputs`. Again, make sure your `pfrom-inputs` have only ONE element.

### CODE REVIEW

**Comment by paweljanicki** [comment]
Looks good to me! Let's go over the merge conflicts in a call. The new approach to validation that you introduced, after the merge conflicts are resolved, let's modify wizardService to handle that, and maybe other stage components if needed.

File: src/dealWizard/validation.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import { ControllerValidateResult, Rule, ValidationController, ValidationRules } from "aurelia-validation";
+import { IProposalLead } from "../entities/Deal";
+
+export const proposalLeadValidationRules = ValidationRules
+  .ensure<IProposalLead, string>(proposalLead => proposalLead.address)
+  .satisfiesRule("isETHAddress")
```
Comment (hiaux0) [inline src/dealWizard/validation.ts:6]: Wdyt about creating an enum or const for `isETHAddress`?
Can then have some kind of map for all the custom validation rules

File: src/dealWizard/components/proposalLeadSection/proposalLeadSection.html [hunk 2]
```diff
@@ -4,23 +4,39 @@
...
+        </pbutton>
       </pform-input>
 
+
       <pform-input
         class="wizardFormInput wizardFormInputNarrow"
-        label="Contact E-mail (optional)">
+        label="Contact E-mail (optional)"
+        validation-message.bind="errors.email"
+        validation-state.bind="errors.email ? 'error' : ''"
```
[the remark this answers is not in the document]
Comment (IonelLupu) [reply src/dealWizard/components/proposalLeadSection/proposalLeadSection.html:38]: I thought about that but the validation-state can also be `warning` or `validating` and not only `error`.

**Comment by giladt** [comment]
Is it possible to resolve error messages after the textbox is cleared or a valid value is entered?
I think it might provide a better UX.
input errors

**Comment by IonelLupu** [comment]
> Is it possible to resolve error messages after the textbox is cleared or a valid value is entered? I think it might provide a better UX. input errors
 @giladt 
@hiaux0 was also mentioning this. I tried to do so but it seems is a lot more complicated that I thought because Aurelia has a weird way of doing validation. We need to investigate this and implement it on every form we have for better UX.

**Review by dkent600 — CHANGES_REQUESTED** [review]
No requests, just to get this out of my requested-review list

**Review by giladt — CHANGES_REQUESTED** [review]
Having a look at the preview, I've noticed 2 issues you might like to address:
1. EMail validation- entering an email with the format `xxx@yyy` is valid.
2. Stage one (if not already on your agenda), can you please apply these changes there too, so that the validations will be resolved upon input changes?

File: src/dealWizard/components/proposalLeadSection/proposalLeadSection.html [hunk 3]
```diff
@@ -4,24 +4,41 @@
       <h2 class="heading heading3 title">Proposal Lead</h2>
       <p class="subtext">
         Provide the Ethereum address for the proposal lead who will gain admin rights of the deal.
-        <br />
-        <br />
-        Proposal leads will be responsible for editing the deal during negotiation, initiating the funding phase once a deal is approved, and executing the token swap. You don’t need to be part of a DAO to be the lead proposal.
+        <br/>
```
Comment (dkent600) [inline src/dealWizard/components/proposalLeadSection/proposalLeadSection.html:7]: Ii think this should be done with a class + margin or padding, rather then brs

File: src/dealWizard/components/proposalLeadSection/proposalLeadSection.ts [hunk 4]
```diff
@@ -1,9 +1,38 @@
...
+  ethAddress?: string;
+
+  private accountSubscription: Subscription;
+
+  constructor(
+    private eventAggregator: EventAggregator,
+    private ethereumService: EthereumService,
+  ) {
+    this.ethAddress = this.ethereumService.defaultAccountAddress;
+    this.accountSubscription = this.eventAggregator.subscribe("Network.Changed.Account", address => {
```
Comment (dkent600) [inline src/dealWizard/components/proposalLeadSection/proposalLeadSection.ts:25]: should be located in attached, to match the detached

File: src/dealWizard/components/proposalLeadSection/proposalLeadSection.ts [hunk 5]
```diff
@@ -1,9 +1,38 @@
...
+      this.ethAddress = address;
+    });
+
+  }
+
+  async connectToWallet() {
+    await this.ethereumService.connect();
```
Comment (dkent600) [inline src/dealWizard/components/proposalLeadSection/proposalLeadSection.ts:32]: no need for await here.  Doesn't behave as you might expect, anyways

File: src/dealWizard/openProposalWizard/openProposalProposalLeadStage/openProposalProposalLeadStage.ts [hunk 6]
```diff
@@ -1,33 +1,39 @@
...
 @autoinject
 export class OpenProposalProposalLeadStage implements IBaseWizardStage {
   public wizardManager: any;
   public wizardState: IWizardState<IDealRegistrationTokenSwap>;
   public errors: WizardErrors<IProposalLead> = {};
+  public form: ValidationController;
 
-  constructor(public wizardService: WizardService) {}
+  constructor(public wizardService: WizardService, validationFactory: ValidationControllerFactory) {
+    this.form = validationFactory.createForCurrentScope();
```
Comment (dkent600) [inline src/dealWizard/openProposalWizard/openProposalProposalLeadStage/openProposalProposalLeadStage.ts:18]: can we make this automatically happen for all stages by moving it into the WizardService?

File: src/dealWizard/openProposalWizard/openProposalProposalLeadStage/openProposalProposalLeadStage.ts [hunk 7]
```diff
@@ -1,33 +1,39 @@
...
   }
 
   attached(): void {
     this.wizardState = this.wizardService.getWizardState(this.wizardManager);
-    this.wizardService.registerStageValidateFunction(this.wizardManager, this.validate.bind(this));
-  }
-
-  validate(): boolean {
-    this.errors = {};
+    this.form.addObject(this.wizardState.registrationData.proposalLead, proposalLeadValidationRules.rules);
```
Comment (dkent600) [inline src/dealWizard/openProposalWizard/openProposalProposalLeadStage/openProposalProposalLeadStage.ts:29]: I would hate to have to jump though all these hoops for every state validation function.  Is possible to make the hoops reusable somehow?

File: src/dealWizard/components/proposalLeadSection/proposalLeadSection.ts [hunk 8]
```diff
@@ -1,9 +1,38 @@
 import { bindable } from "aurelia-typed-observable-plugin";
 import { IProposalLead } from "entities/DealRegistrationTokenSwap";
 import { WizardErrors } from "services/WizardService";
+import "./proposalLeadSection.scss";
```
Comment (dkent600) [inline src/dealWizard/components/proposalLeadSection/proposalLeadSection.ts:4]: Why have proposalLeadSection if only used once?

**Review by dkent600 — CHANGES_REQUESTED** [review]
can we just have one proposalLeadStage component?  Our pattern is to use the wizardType to customize each tage according to the context.  SHould go under the stages folder.  Check with @paweljanicki on this.  we should delete the open and partnered stages stuff, which is deprecated.

File: src/dealWizard/components/proposalLeadSection/proposalLeadSection.ts (hunk 8 above)
Comment (dkent600) [reply src/dealWizard/components/proposalLeadSection/proposalLeadSection.ts:4]: ok, this can be removed once you have just a single stage component

Comment (IonelLupu) [reply src/dealWizard/components/proposalLeadSection/proposalLeadSection.ts:4]: This component is actually used three times throughout the app. Or what do you mean?
image

File: src/dealWizard/openProposalWizard/openProposalProposalLeadStage/openProposalProposalLeadStage.ts (hunk 7 above)
Comment (IonelLupu) [reply src/dealWizard/openProposalWizard/openProposalProposalLeadStage/openProposalProposalLeadStage.ts:29]: what do you mean?

File: src/dealWizard/components/proposalLeadSection/proposalLeadSection.ts (hunk 5 above)
Comment (IonelLupu) [reply src/dealWizard/components/proposalLeadSection/proposalLeadSection.ts:32]: It's good practice to await promises usually. (and my IDE complains about it. I just wanted to get rid of the warning actually 🤣 )

What do you mean about "Doesn't behave as you might expect" ?

File: src/dealWizard/openProposalWizard/openProposalProposalLeadStage/openProposalProposalLeadStage.ts (hunk 6 above)
Comment (IonelLupu) [reply src/dealWizard/openProposalWizard/openProposalProposalLeadStage/openProposalProposalLeadStage.ts:18]: I moved this in the proposalLeadSection component

File: src/dealWizard/components/proposalLeadSection/proposalLeadSection.html (hunk 3 above)
Comment (IonelLupu) [reply src/dealWizard/components/proposalLeadSection/proposalLeadSection.html]: I say those while editing (were there before) and I thought about the same thing. I added design system spacing classes and used those instead

**Comment by IonelLupu** [comment]
@giladt proper email validation and stage 1 validation added

**Review by dkent600 — CHANGES_REQUESTED** [review]
Just resetting the review request

File: src/main.ts [hunk 9]
```diff
@@ -17,13 +17,15 @@ import { TokenService } from "services/TokenService";
 import { CeramicServiceMock } from "services/CeramicServiceMock";
 import { DealTokenSwap } from "entities/DealTokenSwap";
 import { IDataSourceDeals } from "services/DataSourceDealsTypes";
+import "./validation";
```
Comment (dkent600) [inline src/main.ts:20]: do we need this line?

File: src/resources/elements/primeDesignSystem/pform-input/pform-input.ts [hunk 10]
```diff
@@ -15,46 +15,49 @@ export class PFormInput {
...
 
+  validationStateChanged(newValue?: ValidationState) {
+    if (this.inputReference) {
+      this.inputReference.validationState = newValue;
+    }
+  }
+
   private limitInputCharacterLength() {
-    const inputValueObserver = this.bindingEngine.propertyObserver(this.input, "value");
+    const inputValueObserver = this.bindingEngine.propertyObserver(this.inputReference, "value");
```
Comment (dkent600) [inline src/resources/elements/primeDesignSystem/pform-input/pform-input.ts:53]: FYI we have AureliaHelperService for doing this.  no worries.

File: src/resources/elements/primeDesignSystem/pform-input/pform-input.ts [hunk 11]
```diff
@@ -15,46 +15,49 @@ export class PFormInput {
   @bindable.string helperMessage = "";
   @bindable validationMessage = "";
   @bindable.string @observable validationState?: ValidationState;
+  @bindable inputReference;
```
Comment (dkent600) [inline src/resources/elements/primeDesignSystem/pform-input/pform-input.ts:18]: ` @bindable.string` is alreadly observable.  No need for `@observable`.

Comment (dkent600) [reply src/resources/elements/primeDesignSystem/pform-input/pform-input.ts:18]: would be good if you can comment what inputReference is about.

File: src/validation.ts [hunk 12]
```diff
@@ -0,0 +1,20 @@
+import { ValidationRules } from "aurelia-validation";
```
Comment (dkent600) [inline src/validation.ts:1]: "src" is not an appropriate location.   I recommend a `ValidationService.ts` in services.

File: tsconfig.json [hunk 13]
```diff
@@ -12,7 +12,7 @@
     "sourceMap": true,
     "target": "es2019",
     "lib": [
-      "es2019",
+      "ESNext",
```
Comment (dkent600) [inline tsconfig.json:15]: why?

**Review by dkent600 — CHANGES_REQUESTED** [review]

File: tsconfig.json (hunk 13 above)
Comment (IonelLupu) [reply tsconfig.json:15]: I added this library in order to have access to the latest JS features (like String.replaceAll() )

File: src/validation.ts (hunk 12 above)
Comment (IonelLupu) [reply src/validation.ts:1]: it's not really a service. Is just a file with custom validation rules. Is there a place for such things?

File: src/main.ts (hunk 9 above)
Comment (IonelLupu) [reply src/main.ts:20]: Yes. Aurelia will read this file and register all the custom validation rules from it

File: src/validation.ts (hunk 12 above)
Comment (dkent600) [reply src/validation.ts:1]: I would put it where I suggested.  I think calling it a service is ok.

**Review by dkent600 — APPROVED** [review]

**Pull request merged**